modules/context/context.go

```diff
@@ -345,36 +345,54 @@ func (ctx *Context) RespHeader() http.Header {
 	return ctx.Resp.Header()
 }
 
+type ServeHeaderOptions struct {
+	ContentType        string // defaults to "application/octet-stream"
+	ContentTypeCharset string
+	Disposition        string // defaults to "attachment"
+	Filename           string
+	CacheDuration      time.Duration // defaults to 5 minutes
+}
+
 // SetServeHeaders sets necessary content serve headers
-func (ctx *Context) SetServeHeaders(filename string) {
-	ctx.Resp.Header().Set("Content-Description", "File Transfer")
-	ctx.Resp.Header().Set("Content-Type", "application/octet-stream")
-	ctx.Resp.Header().Set("Content-Disposition", "attachment; filename="+filename)
-	ctx.Resp.Header().Set("Content-Transfer-Encoding", "binary")
-	ctx.Resp.Header().Set("Expires", "0")
-	ctx.Resp.Header().Set("Cache-Control", "must-revalidate")
-	ctx.Resp.Header().Set("Pragma", "public")
-	ctx.Resp.Header().Set("Access-Control-Expose-Headers", "Content-Disposition")
+func (ctx *Context) SetServeHeaders(opts *ServeHeaderOptions) {
+	header := ctx.Resp.Header()
+
+	contentType := "application/octet-stream"
+	if opts.ContentType != "" {
+		if opts.ContentTypeCharset != "" {
+			contentType = opts.ContentType + "; charset=" + strings.ToLower(opts.ContentTypeCharset)
+		} else {
+			contentType = opts.ContentType
+		}
+	}
+	header.Set("Content-Type", contentType)
+	header.Set("X-Content-Type-Options", "nosniff")
+
+	if opts.Filename != "" {
+		disposition := opts.Disposition
+		if disposition == "" {
+			disposition = "attachment"
+		}
+
+		header.Set("Content-Disposition", fmt.Sprintf(`%s; filename="%s"; filename*=UTF-8''%s`, disposition, opts.Filename, url.PathEscape(opts.Filename)))
```

Comment:
It will break the `filename="%s"` if `opts.Filename` contains the quote. I think it needs some sanitizing here.

Reply:
Didn't know `"` is an allowed filename character in unix. Windows prevents renaming a file with `"` in its name.

Reply:
Yup, it's allow in POSIX, quite relax: https://stackoverflow.com/questions/457994/what-characters-should-be-restricted-from-a-unix-file-name

Reply:
I added quoting to the filename according to RFC6266. But the appendix of that RFC lists:

> Avoid including the "\" character in the quoted-string form of the
> filename parameter, as escaping is not implemented by some user
> agents, and "\" can be considered an illegal path character.

So, hope for the best?

> Note that this advice is based upon UA behavior at the time of
> writing, and might be superseded.

🤷‍♂️

Reply:
ps: according to rfc822, I think usually it needs at least two replacements: `\` to `\\` and then `"` to `\"`, otherwise the one replacement will make the original string `\"` become `filename="\\"; ...` which still breaks the value.

(small nit)

Reply:
And modern browser should be able to parse the modern `filename*` first, so as long as the `filename` is not broken, it won't take affect in modern browsers.

Reply:
I hope we never serve a file with the name `\`. 😅

Reply:
> Didn't know " is an allowed filename character in unix

Available characters depends on File system, for example ext4:

> All bytes except [NUL](https://en.wikipedia.org/wiki/Null_character) ('\0') and '/' and the special file names "." and ".." which are not forbidden but are always used for a respective special purpose.